Fix delete of `void*`

Mingw flags the following warnings:
```
PaneButton.cpp: In destructor ‘PaneButton::~PaneButton()’:
PaneButton.cpp:196:10: warning: deleting ‘void*’ is undefined [-Wdelete-incomplete]
   delete internalBtn;
          ^~~~~~~~~~~
PaneButton.cpp:197:10: warning: deleting ‘void*’ is undefined [-Wdelete-incomplete]
   delete internalVtbl;
          ^~~~~~~~~~~~
```

```
PaneReport.cpp: In destructor ‘PaneReport::~PaneReport()’:
PaneReport.cpp:140:9: warning: deleting ‘void*’ is undefined [-Wdelete-incomplete]
  delete internalRpt;
         ^~~~~~~~~~~
PaneReport.cpp:141:9: warning: deleting ‘void*’ is undefined [-Wdelete-incomplete]
  delete internalVtbl;
         ^~~~~~~~~~~~
```

As delete may call a destructor, it needs to know the object type.

----

For PaneButton, the `void*` declarations in the header could be changed to `OP2ButtonVtbl*` and `OP2Button*`. This would require placing a forward declaration in the header for those `struct` types (arguably `class` would be more suitable for the second one). This would also eliminate much of the need for casting within the implementation file.

PaneReport looks similar.

--- a/Source/PaneButton.cpp
+++ b/Source/PaneButton.cpp
@@ -141,7 +141,7 @@ PaneButton::PaneButton()
 	// set up button object
 	internalBtn = new OP2Button;
 
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 	memset(p, 0, sizeof(OP2Button));
 	// save a pointer so the dispatcher function can redirect the call to the right object
 	p->btnPtr = this;
@@ -167,7 +167,7 @@ PaneButton::PaneButton()
 	p->hasButtonText = 4;
 }
 
-PaneButton::PaneButton(void *internalPtr)
+PaneButton::PaneButton(OP2Button *internalPtr)
 {
 	internalBtn = NULL;
 	internalVtbl = NULL;
@@ -183,7 +183,7 @@ PaneButton::PaneButton(void *internalPtr)
 
 	int *p = (int*)internalPtr;
 
-	internalBtn = (OP2Button*)internalPtr;
+	internalBtn = internalPtr;
 	internalVtbl = (void*)*p;
 	isInternalObj = 1;
 }
@@ -230,7 +230,7 @@ void PaneButton::OnClick()
 
 void PaneButton::SetParams(int pixelX, int pixelY, int animId, int normalFrameId, int activeFrameId, int disabledFrameId, char *helpText, char *label)
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return;
@@ -252,7 +252,7 @@ void PaneButton::SetParams(int pixelX, int pixelY, int animId, int normalFrameId
 
 void PaneButton::SetHelpText(char *helpText)
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return;
@@ -262,7 +262,7 @@ void PaneButton::SetHelpText(char *helpText)
 
 char* PaneButton::GetHelpText()
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return (char*)HFLNOTINITED;
@@ -272,7 +272,7 @@ char* PaneButton::GetHelpText()
 
 void PaneButton::SetLabel(char *label)
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return;
@@ -282,7 +282,7 @@ void PaneButton::SetLabel(char *label)
 
 char* PaneButton::GetLabel()
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return (char*)HFLNOTINITED;
@@ -292,7 +292,7 @@ char* PaneButton::GetLabel()
 
 int PaneButton::GetPushedIn()
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return HFLNOTINITED;
@@ -302,7 +302,7 @@ int PaneButton::GetPushedIn()
 
 void PaneButton::SetPushedIn(int boolPushed)
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return;
@@ -315,7 +315,7 @@ void PaneButton::SetPushedIn(int boolPushed)
 
 int PaneButton::GetAcceleratorKey()
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return HFLNOTINITED;
@@ -325,7 +325,7 @@ int PaneButton::GetAcceleratorKey()
 
 void PaneButton::SetAcceleratorKey(int asciiCode)
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return;
@@ -335,7 +335,7 @@ void PaneButton::SetAcceleratorKey(int asciiCode)
 
 RECT* PaneButton::GetBoundingBox()
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return NULL;
@@ -347,13 +347,13 @@ ReportButton::ReportButton()
 {
 }
 
-ReportButton::ReportButton(void *internalPtr) : PaneButton(internalPtr)
+ReportButton::ReportButton(OP2Button *internalPtr) : PaneButton(internalPtr)
 {
 }
 
 PaneReport ReportButton::GetAttachedReport()
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return PaneReport(0);
@@ -363,7 +363,7 @@ PaneReport ReportButton::GetAttachedReport()
 
 void ReportButton::SetAttachedReport(PaneReport *newReport)
 {
-	OP2Button *p = (OP2Button*)internalBtn;
+	OP2Button *p = internalBtn;
 
 	if (!p) // not inited if this is null
 		return;

--- a/Source/PaneButton.cpp
+++ b/Source/PaneButton.cpp
@@ -358,7 +358,7 @@ PaneReport ReportButton::GetAttachedReport()
 	if (!p) // not inited if this is null
 		return PaneReport(0);
 
-	return PaneReport(p->report);
+	return PaneReport((OP2Report*)p->report);
 }
 
 void ReportButton::SetAttachedReport(PaneReport *newReport)

--- a/Source/PaneButton.cpp
+++ b/Source/PaneButton.cpp
@@ -183,7 +183,7 @@ PaneButton::PaneButton(void *internalPtr)
 
 	int *p = (int*)internalPtr;
 
-	internalBtn = internalPtr;
+	internalBtn = (OP2Button*)internalPtr;
 	internalVtbl = (void*)*p;
 	isInternalObj = 1;
 }

--- a/Source/PaneButton.cpp
+++ b/Source/PaneButton.cpp
@@ -40,7 +40,7 @@ struct OP2Button
 	int hasButtonText; // set to 4 if the button has text drawn on it, 0 if not
 	int unknown2[22];
 	OP2ButtonData data;
-	void *report; // all buttons are 'report buttons' but we won't use them that way necessarily
+	OP2Report *report; // all buttons are 'report buttons' but we won't use them that way necessarily
 	int unknown3;
 	int fillingSpace[10]; // maybe more stuff that OP2 depends on here..
 	PaneButton *btnPtr; // our PaneButton
@@ -358,7 +358,7 @@ PaneReport ReportButton::GetAttachedReport()
 	if (!p) // not inited if this is null
 		return PaneReport(0);
 
-	return PaneReport((OP2Report*)p->report);
+	return PaneReport(p->report);
 }
 
 void ReportButton::SetAttachedReport(PaneReport *newReport)

--- a/Source/PaneButton.cpp
+++ b/Source/PaneButton.cpp
@@ -149,7 +149,7 @@ PaneButton::PaneButton()
 	// set up vtbl
 	internalVtbl = new OP2ButtonVtbl;
 	
-	OP2ButtonVtbl *vp = (OP2ButtonVtbl*)internalVtbl;
+	OP2ButtonVtbl *vp = internalVtbl;
 	p->vtbl = vp;
 
 	vp->Destructor = (void* (__fastcall *)(OP2Button*,int,int))(imageBase + 0xA8F0);
@@ -181,10 +181,8 @@ PaneButton::PaneButton(OP2Button *internalPtr)
 		}
 	}
 
-	int *p = (int*)internalPtr;
-
 	internalBtn = internalPtr;
-	internalVtbl = (void*)*p;
+	internalVtbl = internalPtr->vtbl;
 	isInternalObj = 1;
 }
 

--- a/Source/PaneButton.h
+++ b/Source/PaneButton.h
@@ -11,7 +11,7 @@ class PaneButton
 {
 public:
 	PaneButton();
-	PaneButton(void *internalPtr);
+	PaneButton(OP2Button *internalPtr);
 	~PaneButton();
 
 	virtual void Paint(PaneGFXSurface gfxSurface);
@@ -41,7 +41,7 @@ class ReportButton : public PaneButton
 {
 public:
 	ReportButton();
-	ReportButton(void *internalPtr);
+	ReportButton(OP2Button *internalPtr);
 	PaneReport GetAttachedReport();
 	void SetAttachedReport(PaneReport *newReport);
 };

--- a/Source/PaneButton.h
+++ b/Source/PaneButton.h
@@ -3,6 +3,10 @@
 #ifndef _PANEBUTTON_H_
 #define _PANEBUTTON_H_
 
+
+struct OP2Button;
+
+
 class PaneButton
 {
 public:
@@ -29,7 +33,7 @@ public:
 	RECT* GetBoundingBox();
 
 	void *internalVtbl;
-	void *internalBtn;
+	OP2Button *internalBtn;
 	int isInternalObj;
 };
 

--- a/Source/PaneButton.h
+++ b/Source/PaneButton.h
@@ -5,6 +5,7 @@
 
 
 struct OP2Button;
+struct OP2ButtonVtbl;
 
 
 class PaneButton
@@ -32,7 +33,7 @@ public:
 	void SetAcceleratorKey(int asciiCode);
 	RECT* GetBoundingBox();
 
-	void *internalVtbl;
+	OP2ButtonVtbl *internalVtbl;
 	OP2Button *internalBtn;
 	int isInternalObj;
 };

--- a/Source/PaneReport.cpp
+++ b/Source/PaneReport.cpp
@@ -103,7 +103,7 @@ PaneReport::PaneReport()
 	// set up vtbl
 	internalVtbl = new OP2ReportVtbl;
 
-	OP2ReportVtbl *vp = (OP2ReportVtbl*)internalVtbl;
+	OP2ReportVtbl *vp = internalVtbl;
 	p->vtbl = vp;
 
 	// todo: some vtbl entries need changing for reports with lists
@@ -127,10 +127,8 @@ PaneReport::PaneReport(OP2Report *internalPtr)
 	if (!isInited)
 		return;
 
-	int *p = (int*)internalPtr;
-
 	internalRpt = internalPtr;
-	internalVtbl = (void*)*p;
+	internalVtbl = internalPtr->vtbl;
 	isInternalObj = 1;
 }
 

--- a/Source/PaneReport.cpp
+++ b/Source/PaneReport.cpp
@@ -129,7 +129,7 @@ PaneReport::PaneReport(void *internalPtr)
 
 	int *p = (int*)internalPtr;
 
-	internalRpt = internalPtr;
+	internalRpt = (OP2Report*)internalPtr;
 	internalVtbl = (void*)*p;
 	isInternalObj = 1;
 }

--- a/Source/PaneReport.cpp
+++ b/Source/PaneReport.cpp
@@ -95,7 +95,7 @@ PaneReport::PaneReport()
 	// set up report object
 	internalRpt = new OP2Report;
 
-	OP2Report *p = (OP2Report*)internalRpt;
+	OP2Report *p = internalRpt;
 	memset(p, 0, sizeof(OP2Report));
 	// save a pointer so the dispatcher function can redirect the call to the right object
 	p->rptPtr = this;
@@ -118,7 +118,7 @@ PaneReport::PaneReport()
 	vp->GetButtonId = ReportGetButtonIdDispatcher;
 }
 
-PaneReport::PaneReport(void *internalPtr)
+PaneReport::PaneReport(OP2Report *internalPtr)
 {
 	isInternalObj = 0;
 	internalRpt = NULL;
@@ -129,7 +129,7 @@ PaneReport::PaneReport(void *internalPtr)
 
 	int *p = (int*)internalPtr;
 
-	internalRpt = (OP2Report*)internalPtr;
+	internalRpt = internalPtr;
 	internalVtbl = (void*)*p;
 	isInternalObj = 1;
 }
@@ -164,7 +164,7 @@ int PaneReport::GetLinkedButtonId()
 
 void PaneReport::DrawTitle(RECT *updateRect, PaneGFXSurface gfxSurface, char *title)
 {
-	OP2Report *p = (OP2Report*)internalRpt;
+	OP2Report *p = internalRpt;
 
 	if (!p)
 		return;
@@ -179,7 +179,7 @@ void PaneReport::CreateStdButtons(int numButtons, PaneButton *button1, char *but
 	// act as if the user passed parameters for all 4 buttons (the stack will balance out okay)
 	// Performance hit but it will prevent having to deal with passing parameters manually
 
-	OP2Report *p = (OP2Report*)internalRpt;
+	OP2Report *p = internalRpt;
 
 	if (!p)
 		return;
@@ -209,7 +209,7 @@ void PaneReport::CreateStdButtons(int numButtons, PaneButton *button1, char *but
 
 int PaneReport::GetCurrentPage()
 {
-	OP2Report *p = (OP2Report*)internalRpt;
+	OP2Report *p = internalRpt;
 
 	if (!p)
 		return HFLNOTINITED;
@@ -219,7 +219,7 @@ int PaneReport::GetCurrentPage()
 
 void PaneReport::SetCurrentPage(int pageNum)
 {
-	OP2Report *p = (OP2Report*)internalRpt;
+	OP2Report *p = internalRpt;
 
 	if (!p)
 		return;

--- a/Source/PaneReport.h
+++ b/Source/PaneReport.h
@@ -13,7 +13,7 @@ class PaneReport
 {
 public:
 	PaneReport();
-	PaneReport(void *internalPtr);
+	PaneReport(OP2Report *internalPtr);
 	~PaneReport();
 
 	virtual void Update();

--- a/Source/PaneReport.h
+++ b/Source/PaneReport.h
@@ -7,6 +7,7 @@
 // forward declare is needed since the other header isn't included yet
 class PaneButton;
 struct OP2Report;
+struct OP2ReportVtbl;
 
 
 class PaneReport
@@ -28,7 +29,7 @@ public:
 	void SetCurrentPage(int pageNum);
 	// todo: listcontrol stuff in the future
 
-	void *internalVtbl;
+	OP2ReportVtbl *internalVtbl;
 	OP2Report *internalRpt;
 	int isInternalObj;
 };

--- a/Source/PaneReport.h
+++ b/Source/PaneReport.h
@@ -3,8 +3,11 @@
 #ifndef _PANEREPORT_H_
 #define _PANEREPORT_H_
 
+
 // forward declare is needed since the other header isn't included yet
 class PaneButton;
+struct OP2Report;
+
 
 class PaneReport
 {
@@ -26,7 +29,7 @@ public:
 	// todo: listcontrol stuff in the future
 
 	void *internalVtbl;
-	void *internalRpt;
+	OP2Report *internalRpt;
 	int isInternalObj;
 };
 

--- a/Source/StdPanes.cpp
+++ b/Source/StdPanes.cpp
@@ -73,7 +73,7 @@ ReportButton CommandPane::GetReportButton(ReportButtonType which)
 
 	OP2CommandPane *p = (OP2CommandPane*)cmdPane;
 
-	return ReportButton(p->reportButtons[which-1]);
+	return ReportButton((OP2Button*)p->reportButtons[which-1]);
 }
 
 void CommandPane::AddButton(PaneButton *btn)

